gprofiler/spark/spark_collector.py

```diff
@@ -0,0 +1,894 @@
+
+        return _url
+
+    @staticmethod
+    def _join_url_dir(url: str, *args: Any) -> str:
+        """
+        Join a URL with multiple directories
+        """
+        for path in args:
+            url = url.rstrip("/") + "/"
+            url = urljoin(url, path.lstrip("/"))
+
+        return url
+
+
+class SparkSampler(object):
+    """
+    Spark cluster metrics sampler
+    """
+
+    def __init__(
+        self,
+        sample_period: float,
+        storage_dir: str,
+        api_client: Optional[APIClient] = None,
+    ):
+        self._master_address: Optional[str] = None
+        self._spark_mode: Optional[str] = None
+        self._collection_thread: Optional[Thread] = None
+        self._logger = get_logger_adapter(__name__)
+        self._sample_period = sample_period
+        self._stop_event = Event()
+        self._spark_sampler: Optional[SparkCollector] = None
+        self._stop_collection = False
+        self._is_running = False
+        self._storage_dir = storage_dir
+        if self._storage_dir is not None:
+            assert os.path.exists(self._storage_dir) and os.path.isdir(self._storage_dir)
+        else:
+            self._logger.debug("output directory is None. Will add metrics to queue")
+        self._client = api_client
+
+    def _get_yarn_config_path(self, process: psutil.Process) -> str:
+        env = process.environ()
+        if "HADOOP_CONF_DIR" in env:
+            self._logger.debug("Found HADOOP_CONF_DIR variable.", extra={"hadoop_conf_dir": env["HADOOP_CONF_DIR"]})
+            return os.path.join(env["HADOOP_CONF_DIR"], "yarn-site.xml")
+        else:
+            self._logger.info(
+                "Could not find HADOOP_CONF_DIR variable, using default path",
+                extra={"hadoop_conf_dir": os.path.join("/etc/hadoop/conf/", "yarn-site.xml")},
+            )
+            return os.path.join("/etc/hadoop/conf/", "yarn-site.xml")
+
+    @staticmethod
+    def _get_process_root_relative_path(process: psutil.Process, absolute_path: str) -> str:
+        return os.path.join(f"/proc/{process.pid}/root", absolute_path.lstrip(os.sep))
+
+    def _get_yarn_config(self, process: psutil.Process) -> Optional[ET.Element]:
+        config_path = self._get_yarn_config_path(process)
+
+        self._logger.debug("Trying to open yarn config file for reading", extra={"config_path": config_path})
+        try:
+            # resolve config path against process' filesystem root
+            process_relative_config_path = self._get_process_root_relative_path(
+                process, self._get_yarn_config_path(process)
+            )
+            with open(process_relative_config_path, "rb") as conf_file:
+                config_xml_string = conf_file.read()
+            return ET.fromstring(config_xml_string)
+        except FileNotFoundError:
+            return None
+
+    def _get_yarn_config_property(
+        self, process: psutil.Process, requested_property: str, default: Any = None
```

Comment:
```suggestion
        self, process: psutil.Process, requested_property: str, default: Optional[str] = None
```

Per the function's return value

Reply:
Done